### PR TITLE
@#1 | implement variable processor

### PR DESCRIPTION
**Pull request by datphan**
connect #1

### CODE REVIEW

**Comment by hoatle** [comment]
this is how variables works:

config_default.yaml

```
variables:
  node_name_prefix: ${NODE_NAME_PREFIX:-node}
  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}

nodes:
  - _id: "0"
    name: "#{node_name_prefix}-01"
    primary: true
    vm:
      hostname: "#{node_hostname_prefix}-01.local"
```

=> `$ vagrant` should produce final settings:

```
variables:
  node_name_prefix: ${NODE_NAME_PREFIX:-node}
  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}

nodes:
  - _id: "0"
    name: "node-01"
    primary: true
    vm:
      hostname: "node-01.local"
```

and `$ NODE_NAME_PREFIX=teracy NODE_HOSTNAME_PREFIX=teracy vagrant` should produce the following final settings:


```
variables:
  node_name_prefix: ${NODE_NAME_PREFIX:-node}
  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}

nodes:
  - _id: "0"
    name: "#{node_name_prefix}-01"
    primary: true
    vm:
      hostname: "#{node_hostname_prefix}-01.local"
```

=> `$ vagrant` should produce final settings:

```
variables:
  node_name_prefix: ${NODE_NAME_PREFIX:-node}
  node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-node}

nodes:
  - _id: "0"
    name: "teracy-01"
    primary: true
    vm:
      hostname: "teracy-01.local"
```

File: lib/teracy-dev-core/processors/variables.rb [hunk 1]
```diff
@@ -6,7 +6,14 @@ module Processors
     class Variables < TeracyDev::Processors::Processor
 
       def process(settings)
-        @logger.debug("process: #{settings}")
+        @logger.debug("process: #{settings['variables']}")
+
+        # see `settings['variables']` as a template, then load it using `envsubst`
+        variables = `envsubst <<< "#{settings['variables']}"`
```
Comment (hoatle) [inline lib/teracy-dev-core/processors/variables.rb]: note: you can not assume `envsubst` is always available on macOS, Linux, Windows.

**Review by hoatle — CHANGES_REQUESTED** [review]
check my comment

**Comment by datphan** [comment]
fixed  and fixes #2 also

**Comment by datphan** [comment]
unable to have this works: https://ruby-doc.org/core-2.4.0/Binding.html
```
vm:
      hostname: "#{node_hostname_prefix}-01.local"
```
so I have to use this:
```
vm:
      hostname: "%{node_hostname_prefix}-01.local"
```

**Comment by hoatle** [comment]
not working with teracy-dev-essential:

```
There was an error loading a Vagrantfile. The file being loaded
and the error message are shown below. This is usually caused by
a syntax error.

Path: /Users/hoatle/k8s-dev/Vagrantfile
Line number: 0
Message: ArgumentError: malformed format string - %H
```
see: https://github.com/teracyhq-incubator/teracy-dev-essential/blob/develop/config_default.yaml#L15

**Comment by datphan** [comment]
We can use double `%`'s
before:
```
time_format: "%%H:%%M:%%S"
```
after:
```
time_format: "%H:%M:%S"
```

**Review by hoatle — APPROVED** [review]
LGTM

**Pull request merged**